packages/grafana-ui/src/components/Dropdown/Dropdown.tsx

```diff
@@ -27,16 +27,22 @@ export interface Props {
   children: React.ReactElement;
   /** Amount in pixels to nudge the dropdown vertically and horizontally, respectively. */
   offset?: [number, number];
-  onVisibleChange?: (state: boolean) => void;
+  onVisibleChange?: VisibilityChangeHandler;
 }
 
+type VisibilityChangeHandler = (state: boolean) => void | undefined;
+
 export const Dropdown = React.memo(({ children, overlay, placement, offset, onVisibleChange }: Props) => {
   const [show, setShow] = useState(false);
   const transitionRef = useRef(null);
+  const onVisibilityChangeRef = useRef<VisibilityChangeHandler>();
+  onVisibilityChangeRef.current = onVisibleChange;
 
   useEffect(() => {
-    onVisibleChange?.(show);
-  }, [onVisibleChange, show]);
+    if (onVisibilityChangeRef.current) {
+      onVisibilityChangeRef.current(show);
+    }
+  }, [show]);
```

Comment:
Instead of all this workaround stuff, I think the function should be explicitly called when the action happens.

> **Ask yourself why this code needs to run. Use Effects only for code that should run because the component was displayed to the user.** In this example, the notification should appear because the user pressed the button, not because the page was displayed! Delete the Effect and put the shared logic into a function called from both event handlers

https://react.dev/learn/you-might-not-need-an-effect#sharing-logic-between-event-handlers


Down in useFloating, change `onOpenChange` to a handler that both sets the state (calls `setShow`), and calls `onVisibilityChange`.


```ts
const handleOpenChange = useCallback((newState: boolean) => {
  setShow(newState);
  onVisibleChange(newState);
}, [onVisibleChange]);

...

const { context, refs, floatingStyles } = useFloating({
  ...
  onOpenChange: handleOpenChange,
});
```

Reply:
@joshhunt ah, of course. pushed an update